src/Microsoft.AspNet.Mvc.Core/Filters/ResponseCacheAttribute.cs

```diff
@@ -38,97 +38,85 @@ public int Duration
         /// <summary>
         /// Gets or sets the location where the data from a particular URL must be cached.
         /// </summary>
-        public ResponseCacheLocation Location { get; set; }
+        public ResponseCacheLocation Location
+        {
+            get
+            {
+                return _location ?? ResponseCacheLocation.Any;
+            }
+            set
+            {
+                _location = value;
+            }
+        }
 
         /// <summary>
         /// Gets or sets the value which determines whether the data should be stored or not.
-        /// When set to true, it sets "Cache-control" header to "no-store".
+        /// When set to <see langword="true"/>, it sets "Cache-control" header to "no-store".
         /// Ignores the "Location" parameter for values other than "None".
         /// Ignores the "duration" parameter.
         /// </summary>
-        public bool NoStore { get; set; }
+        public bool NoStore
+        {
+            get
+            {
+                return _noStore ?? false;
+            }
+            set
+            {
+                _noStore = value;
+            }
+        }
 
         /// <summary>
         /// Gets or sets the value for the Vary response header.
         /// </summary>
         public string VaryByHeader { get; set; }
 
-        // <inheritdoc />
-        public override void OnActionExecuting([NotNull] ActionExecutingContext context)
-        {
-            // If there are more filters which can override the values written by this filter,
-            // then skip execution of this filter.
-            if (IsOverridden(context))
-            {
-                return;
-            }
-
-            var headers = context.HttpContext.Response.Headers;
-
-            // Clear all headers
-            headers.Remove("Vary");
-            headers.Remove("Cache-control");
-            headers.Remove("Pragma");
+        /// <summary>
+        /// Gets or sets the value of the Cache Profile.
+        /// </summary>
+        public string CacheProfileName { get; set; }
 
-            if (!string.IsNullOrEmpty(VaryByHeader))
-            {
-                headers.Set("Vary", VaryByHeader);
-            }
+        public IFilter CreateInstance([NotNull] IServiceProvider serviceProvider)
+        {
+            var optionsAccessor = serviceProvider.GetRequiredService<IOptions<MvcOptions>>();
 
-            if (NoStore)
+            CacheProfile selectedProfile = null;
+            if (!string.IsNullOrEmpty(CacheProfileName))
             {
-                headers.Set("Cache-control", "no-store");
+                selectedProfile = optionsAccessor.Options.CacheProfiles.FirstOrDefault(
```

Comment:
shouldnt it be SingleOrDefault? can you have more than one cache profiles registered with the same name?